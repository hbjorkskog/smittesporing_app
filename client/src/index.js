// @flow

import ReactDOM from 'react-dom';
import * as React from 'react';
import { Component } from 'react-simplified';
import { HashRouter, Route } from 'react-router-dom';
import { HeaderCard, Card, Alert, Form, Row, Column, Button } from './widgets';
import nameService, { type Name } from './name-service';
import { createHashHistory } from 'history';

const history = createHashHistory(); // Use history.push(...) to programmatically change path, for instance after successfully adding a name

class Header extends Component {
  render() {
    return (
      <div className="headerElement">
        <HeaderCard title="Covid-19 contact tracing"></HeaderCard>
      </div>
    );
  }
}
class RegisterName extends Component {
  name = '';
  phone = '';

  render() {
    return (
      <div className="register">
        <Card title="Register visit">
          <Row className="nameRow">
            <Column width={5}>
              <Form.Label>Name:</Form.Label>
            </Column>
            <Column width={10}>
              <Form.Input
                type="text"
                value={this.name}
                onChange={(event) => (this.name = event.currentTarget.value)}
                placeholder="Insert your name here"
              />
            </Column>
          </Row>
          <Row className="phoneRow">
            <Column width={5}>
              <Form.Label>Phone number:</Form.Label>
            </Column>
            <Column width={10}>
              <Form.Input
                type="text"
                value={this.phone}
                onChange={(event) => (this.phone = event.currentTarget.value)}
                placeholder="Insert your phone number here"
              />
            </Column>
          </Row>
          <Button.Success
            onClick={() => {
              nameService
                .create(this.name, this.phone)
                .then(() => history.push('/'))
                .catch((error: Error) => Alert.danger('Registation failed: ' + error.message));
              /* Alert.success('Registration successful'); */
            }}
          >
            Register
          </Button.Success>
        </Card>
      </div>
    );
  }
}

class About extends Component {
  render() {
    return (
      <div className="aboutApp">
        <Card title="About this application">
          <Row className="aboutRow">
            <Column>
              This is an application for Covid-19 contact tracing. The information will be deleted
              after 10 days and will not be used for any other purpose than contact tracing.
            </Column>
          </Row>
          <Row>
            <Column>
              <hr></hr>
            </Column>
          </Row>
          <Row>
            <Column className="thankYou">
              Thank you for helping us do our part in fighting this pandemic.
            </Column>
          </Row>
        </Card>
      </div>
    );
  }
}

const root = document.getElementById('root');
if (root)
  ReactDOM.render(
    <HashRouter>
      <div>
        <Alert />
        <Header />
        <RegisterName />
        <About />
      </div>
    </HashRouter>,
    root
  );
