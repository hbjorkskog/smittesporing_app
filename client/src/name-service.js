// @flow
import axios from 'axios';

axios.defaults.baseURL = 'http://localhost:3000/api/v2';

export type Name = {
  id: number,
  name: string,
  phone: string,
};

class NameService {
  /* The functions for get(), getAll() and delete() will not be used in this application so they are commented out. However, they can be used in case there is need for a  "Show all names on list" page later so they are left in the file. If they are used later, it would also require authentication.

  // Get specific name
  get(id: number) {
    return axios.get<Name>('/names/' + id).then((response) => response.data);
  }
  // Get all names
  getAll() {
    return axios.get<Name[]>('/names').then((response) => response.data);
  } */

  /**
   * Create new name.
   */
  create(name: string, phone: string) {
    return axios
      .post<{}, { id: number }>('/names', {
        name: name,
        phone: phone,
      })
      .then((response) => response.data.id);
  }

  /*
  // Delete name
  delete(id: number) {
    return axios.delete<void>('/names/' + id).then((response) => response.data);
  } */
}

const nameService = new NameService();
export default nameService;
