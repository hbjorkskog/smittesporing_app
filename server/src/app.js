// @flow

import express from 'express';
import nameRouter from './name-router';

/**
 * Express application.
 */
const app: express$Application<> = express();

app.use(express.json());

// Since API is not compatible with v1, API version is increased to v2
app.use('/api/v2', nameRouter);

export default app;
