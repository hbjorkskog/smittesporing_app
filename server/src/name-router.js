// @flow
import express from 'express';
import nameService, { type NameList } from './name-service';

/**
 * Express router containing nameList methods.
 */
const router: express$Router<> = express.Router();

router.get('/nameList', (request, response) => {
  nameService
    .getAll()
    .then((rows) => response.send(rows))
    .catch((error: Error) => response.status(500).send(error));
});

router.get('/nameList/:id', (request, response) => {
  const id = Number(request.params.id);
  nameService
    .get(id)
    .then((name) => (name ? response.send(name) : response.status(404).send('Task not found')))
    .catch((error: Error) => response.status(500).send(error));
});

// Example request body: { name: "Kari Nordmann", phone: 56565656, time_created:... now() }
// Example response body: { id: 4 }
router.post('/nameList', (request, response) => {
  const data = request.body;
  if (
    data &&
    typeof data.name == 'string' &&
    data.name.length != 0 &&
    typeof data.phone == 'number' &&
    data.name.length != 0 &&
    typeof data.time_created == 'number'
  )
    nameService
      .create(data.name, data.phone, data.time_created)
      .then((id) => response.send({ id: id }))
      .catch((error: Error) => response.status(500).send(error));
  else response.status(400).send('Missing name properties');
});

router.delete('/nameList/:id', (request, response) => {
  nameService
    .delete(Number(request.params.id))
    .then((result) => response.send())
    .catch((error: Error) => response.status(500).send(error));
});

export default router;
