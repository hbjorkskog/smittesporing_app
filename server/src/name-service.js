// @flow

import pool from './mysql-pool';

export type NameList = {
  name: string,
  phone: number,
  time_created: number,
};

class NameService {
  /**
   * Get name with given id.
   */
  get(id: number) {
    return new Promise<?NameList>((resolve, reject) => {
      pool.query('SELECT * FROM NameList WHERE id = ?', [id], (error, results: NameList[]) => {
        if (error) return reject(error);

        resolve(results[0]);
      });
    });
  }

  /**
   * Get all names.
   */
  getAll() {
    return new Promise<NameList[]>((resolve, reject) => {
      pool.query('SELECT * FROM NameList', (error, results) => {
        if (error) return reject(error);

        resolve(results);
      });
    });
  }

  /* SQL Query for create row: 
    INSERT INTO NameList SET name='Anne', phone=56565656, time_created=NOW()
 */

  /**
   * Create new name with phone and timestamp for when created.
   */
  create(name: string, phone: number, time_created: number) {
    return new Promise<number>((resolve, reject) => {
      pool.query(
        'INSERT INTO NameList SET name=?, phone=?, time_created=NOW()',
        [name, phone, time_created],
        (error, results) => {
          if (error) return reject(error);
          if (!results.insertId) return reject(new Error('No row inserted'));

          resolve(Number(results.insertId));
        }
      );
    });
  }

  /**
   * Delete name with given id.
   */
  delete(id: number) {
    return new Promise<void>((resolve, reject) => {
      pool.query('DELETE FROM NameList WHERE id = ?', [id], (error, results) => {
        if (error) return reject(error);
        if (!results.affectedRows) reject(new Error('No row deleted'));

        resolve();
      });
    });
  }
}

const nameService = new NameService();
export default nameService;
