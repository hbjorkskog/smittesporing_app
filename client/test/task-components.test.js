// @flow

import * as React from 'react';
import { TaskList, TaskDetails, TaskEdit, TaskNew } from '../src/task-components';
import { type Task } from '../src/task-service';
import { shallow } from 'enzyme';
import { Form, Button, Column } from '../src/widgets';
import { NavLink } from 'react-router-dom';

jest.mock('../src/task-service', () => {
  class TaskService {
    getAll() {
      return Promise.resolve([
        { id: 1, title: 'Les leksjon', description: 'Les nøye', done: true },
        { id: 2, title: 'Møt opp på forelesning', description: 'I tide', done: false },
        { id: 3, title: 'Gjør øving', description: 'Før fristen', done: false },
      ]);
    }

    get(id: number) {
      return Promise.resolve({ id: 1, title: 'Les leksjon', description: 'Les nøye', done: true });
    }

    create(title: string) {
      return Promise.resolve(4); // Same as: return new Promise((resolve) => resolve(4));
    }

    update(task: Task) {
      return Promise.resolve();
    }

    delete(id: number) {
      return Promise.resolve();
    }
  }
  return new TaskService();
});

describe('Task component tests', () => {
  test('TaskList draws correctly', (done) => {
    const wrapper = shallow(<TaskList />);

    // Wait for events to complete
    setTimeout(() => {
      expect(
        wrapper.containsAllMatchingElements([
          <NavLink to="/tasks/1">Les leksjon</NavLink>,
          <NavLink to="/tasks/2">Møt opp på forelesning</NavLink>,
          <NavLink to="/tasks/3">Gjør øving</NavLink>,
        ])
      ).toEqual(true);
      done();
    });
  });

  test('TaskDetails draws correctly', (done) => {
    const wrapper = shallow(<TaskDetails match={{ params: { id: 1 } }} />);

    setTimeout(() => {
      expect(
        wrapper.containsAllMatchingElements([
          <Column>Les leksjon</Column>,
          <Column>Les nøye</Column>,
          // $FlowExpectedError
          <Form.Checkbox checked={true} />,
        ])
      ).toEqual(true);
      done();
    });
  });

  test('TaskDetails draws correctly (using snapshot)', (done) => {
    const wrapper = shallow(<TaskDetails match={{ params: { id: 1 } }} />);

    setTimeout(() => {
      expect(wrapper).toMatchSnapshot();
      done();
    });
  });

  test('TaskDetails correctly sets location on edit', (done) => {
    const wrapper = shallow(<TaskDetails match={{ params: { id: 1 } }} />);

    setTimeout(() => {
      expect(
        wrapper.containsAllMatchingElements([
          <Column>Les leksjon</Column>,
          <Column>Les nøye</Column>,
          // $FlowExpectedError
          <Form.Checkbox checked={true} />,
        ])
      ).toEqual(true);

      wrapper.find(Button.Success).simulate('click');

      setTimeout(() => {
        expect(location.hash).toEqual('#/tasks/1/edit');
        done();
      });
    });
  });

  test('TaskEdit draws correctly', (done) => {
    const wrapper = shallow(<TaskEdit match={{ params: { id: 1 } }} />);

    setTimeout(() => {
      expect(
        wrapper.containsAllMatchingElements([
          // $FlowExpectedError
          <Form.Input value="Les leksjon" />,
          // $FlowExpectedError
          <Form.TextArea value="Les nøye" />,
          // $FlowExpectedError
          <Form.Checkbox checked={true} />,
        ])
      ).toEqual(true);
      done();
    });
  });

  test('TaskEdit correctly sets location on save', (done) => {
    const wrapper = shallow(<TaskEdit match={{ params: { id: 1 } }} />);

    setTimeout(() => {
      expect(
        wrapper.containsAllMatchingElements([
          // $FlowExpectedError
          <Form.Input value="Les leksjon" />,
          // $FlowExpectedError
          <Form.TextArea value="Les nøye" />,
          // $FlowExpectedError
          <Form.Checkbox checked={true} />,
        ])
      ).toEqual(true);

      wrapper
        .find(Form.Input)
        .simulate('change', { currentTarget: { value: 'Les leksjon igjen' } });
      // $FlowExpectedError
      expect(wrapper.containsMatchingElement(<Form.Input value="Les leksjon igjen" />)).toEqual(
        true
      );

      wrapper
        .find(Form.TextArea)
        .simulate('change', { currentTarget: { value: 'Les ekstra nøye' } });
      // $FlowExpectedError
      expect(wrapper.containsMatchingElement(<Form.TextArea value="Les ekstra nøye" />)).toEqual(
        true
      );

      wrapper.find(Button.Success).simulate('click');

      setTimeout(() => {
        expect(location.hash).toEqual('#/tasks/1');
        done();
      });
    });
  });

  test('TaskEdit correctly sets location on delete', (done) => {
    const wrapper = shallow(<TaskEdit match={{ params: { id: 1 } }} />);

    setTimeout(() => {
      expect(
        wrapper.containsAllMatchingElements([
          // $FlowExpectedError
          <Form.Input value="Les leksjon" />,
          // $FlowExpectedError
          <Form.TextArea value="Les nøye" />,
          // $FlowExpectedError
          <Form.Checkbox checked={true} />,
        ])
      ).toEqual(true);

      wrapper.find(Button.Danger).simulate('click');

      setTimeout(() => {
        expect(location.hash).toEqual('#/tasks');
        done();
      });
    });
  });

  test('TaskNew correctly sets location on create', (done) => {
    const wrapper = shallow(<TaskNew />);

    wrapper.find(Form.Input).simulate('change', { currentTarget: { value: 'Kaffepause' } });
    // $FlowExpectedError
    expect(wrapper.containsMatchingElement(<Form.Input value="Kaffepause" />)).toEqual(true);

    wrapper.find(Button.Success).simulate('click');
    // Wait for events to complete
    setTimeout(() => {
      expect(location.hash).toEqual('#/tasks/4');
      done();
    });
  });
});