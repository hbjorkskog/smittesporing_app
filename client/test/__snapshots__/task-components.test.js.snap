// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task component tests TaskEdit draws correctly 1`] = `
<Fragment>
  <Card
    title="Edit task"
  >
    <Row>
      <Column
        width={2}
      >
        <FormLabel>
          Title:
        </FormLabel>
      </Column>
      <Column>
        <FormInput
          onChange={[Function]}
          type="text"
        />
      </Column>
    </Row>
    <Row>
      <Column
        width={2}
      >
        <FormLabel>
          Description:
        </FormLabel>
      </Column>
      <Column>
        <FormTextarea
          onChange={[Function]}
          rows={10}
        />
      </Column>
    </Row>
    <Row>
      <Column
        width={2}
      >
        Done:
      </Column>
      <Column>
        <FormCheckbox
          onChange={[Function]}
        />
      </Column>
    </Row>
  </Card>
  <Row>
    <Column>
      <ButtonSuccess
        onClick={[Function]}
      >
        Save
      </ButtonSuccess>
    </Column>
    <Column
      right={true}
    >
      <ButtonDanger
        onClick={[Function]}
      >
        Delete
      </ButtonDanger>
    </Column>
  </Row>
</Fragment>
`;

exports[`Task component tests Test TaskDetails using snapshot 1`] = `
<Fragment>
  <Card
    title="Task"
  >
    <Row>
      <Column
        width={2}
      >
        Title:
      </Column>
      <Column />
    </Row>
    <Row>
      <Column
        width={2}
      >
        Description:
      </Column>
      <Column />
    </Row>
    <Row>
      <Column
        width={2}
      >
        Done:
      </Column>
      <Column>
        <FormCheckbox
          checked={false}
          disabled={true}
          onChange={[Function]}
        />
      </Column>
    </Row>
  </Card>
  <ButtonSuccess
    onClick={[Function]}
  >
    Edit
  </ButtonSuccess>
</Fragment>
`;
